lib/src/import_cache.dart

```diff
@@ -206,18 +207,17 @@ final class ImportCache {
   /// that result is cacheable at all.
   (CanonicalizeResult?, bool cacheable) _canonicalize(
       Importer importer, Uri url, Uri? baseUrl, bool forImport) {
-    var canonicalize = forImport
-        ? () => inImportRule(() => importer.canonicalize(url))
-        : () => importer.canonicalize(url);
-
-    var passContainingUrl = baseUrl != null &&
-        (url.scheme == '' || importer.isNonCanonicalScheme(url.scheme));
-    var result =
-        withContainingUrl(passContainingUrl ? baseUrl : null, canonicalize);
-
-    // TODO(sass/dart-sass#2208): Determine whether the containing URL was
-    // _actually_ accessed rather than assuming it was.
-    var cacheable = !passContainingUrl || importer is FilesystemImporter;
+    var canonicalizeContext = CanonicalizeContext(
+        baseUrl != null &&
+                (url.scheme == '' || importer.isNonCanonicalScheme(url.scheme))
+            ? baseUrl
+            : null,
+        forImport);
+
+    var result = withCanonicalizeContext(
+        canonicalizeContext, () => importer.canonicalize(url));
+
+    var cacheable = !canonicalizeContext.wasContainingUrlAccessed;
```

Comment:
Consider treating the containing URL as not accessed if `containingUrlAvailable: url.scheme != '' && !importer.isNonCanonicalScheme(url.scheme)`, since in that case it's not even accessible to the importer. That would make us a little more robust against importers that access `containingUrl` more aggressively than they should.